helsingborg-stad/app-mitt-helsingborg
# Feat/image viewer

## Explain the changes you’ve made

Adds an ImageViewer component, meant for displaying uploaded images, in order to summarize to the user what he has added. 
Also refactors the existing ImageUploader, and improves its design and functionality. 

## Explain your solution

I create a new component, called ImageDisplay, which is responsible for displaying a list of images, and handle removing them. This is used in both the ImageUploader and ImageViewer, so that they automatically have the same functionality and design for this part of their functionality. The ImageDisplay has the added behaviour that the user can click an image to open it in a modal and zoom in on it etc., to inspect the content. 

The ImageViewer is a small wrapper around the ImageDisplay, which lets us summarize images from multiple uploaders in the same place. For example if the user should upload bank statements in one place, and receipts in another, this lets us show all the images in one list, if we want. 

The ImageUploader is refactored in two ways: firstly to use the new ImageDisplay for displaying images. Secondly, I remove the use of an internal useState, and instead just use the passed value prop and the onChange method. This is just more properly functional and better, and it avoids issues I was having with statechanges interrupting animation. 

One thing that is not yet done (I'll make it a separate task) is to implement the functionality for deleting files from the S3 bucket via the new attachment/delete endpoint. 

Other things that we might need as well is the functionality for resuming a cancelled form with attached files. This will need more functionality for handling images taken with the camera and not stored in local storage on the phone. 

## How to test the changes?

Checkout this branch, and run yarn to install one new dependency. Then, in the dev-environment there is a new form called Test Form for Image Viewer component, which can be accessed through the dev-screen. This has some image uploaders as well as imageViewers to test the various features. So test uploading some images using both the uploaders, and check that they show up correctly in the image viewer, that you can tap the images to open them, that it works on the next page as well, and that the remove functionality works, etc. 

## Was this feature tested in the following environments?
- [] Storybook on a iOS device/simulator.
- [] Building the Application on a iOS device/simulator.

## Screenshots (optional)
![imageViewer](https://user-images.githubusercontent.com/7421890/107529274-fdbc5780-6bba-11eb-9536-bb3371adabcd.png)
